art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -283,117 +328,169 @@ def create_model(
-        dataloader_test = torch.utils.data.DataLoader(dataset_test, batch_size=batch_size)
-
-        for epoch in trange(epochs):
-            running_loss = 0.0
-            total = 0
-            accuracy = 0
-            for _, data in enumerate(dataloader_train, 0):
-                inputs, labels = data
-                optimizer.zero_grad()
-                # forward + backward + optimize
-                outputs = model(inputs)
-                loss = loss_fn(outputs, labels)
-                loss.backward()
-                optimizer.step()
-                _, predicted = torch.max(outputs.data, 1)
-                total += labels.size(0)
-                accuracy += (predicted == labels).sum().item()
-                running_loss += loss.item()
-            if (epoch % 5 == 0) or epoch == (epochs - 1):
-                train_accuracy = 100 * accuracy / total
-                logger.info("Epoch %d train accuracy: %f", epoch, train_accuracy)
-        test_accuracy = self.test_accuracy(model, dataloader_test)
-        logger.info("Final test accuracy: %f", test_accuracy)
-        return model, loss_fn, optimizer
-
-    @classmethod
-    def test_accuracy(cls, model: "torch.nn.Module", test_loader: "torch.utils.data.dataloader.DataLoader") -> float:
-        """
-        Calculates test accuracy on trained model
+        from art.estimators.classification.pytorch import PyTorchClassifier
+        from art.estimators.classification import TensorFlowV2Classifier
+        
+        
+        if isinstance(self.substitute_classifier, PyTorchClassifier):
+            import torch
+            import torch.nn as nn
+            import torchvision
+
+            loss_fn = nn.CrossEntropyLoss()
+            model = torchvision.models.ResNet(torchvision.models.resnet.BasicBlock, [2, 2, 2, 2], num_classes=num_classes)
+            optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9, weight_decay=5e-4, nesterov=True)
+            model_new = PyTorchClassifier(model,input_shape=x_train.shape[1:], loss=loss_fn, optimizer=optimizer, nb_classes=num_classes, clip_values=self.substitute_classifier.clip_values, preprocessing=(self.substitute_classifier.preprocessing.mean,self.substitute_classifier.preprocessing.std))
+            model_new.fit(x_train, y_train, batch_size=128, nb_epochs=epochs,verbose=1)
+            predictions = model_new.predict(x_test)
+            accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)
+            print("Accuracy of substitute model after retraining: {}%".format(accuracy * 100))
+        
+        elif isinstance(self.substitute_classifier, TensorFlowV2Classifier):
+            import tensorflow as tf
+            from tqdm.keras import TqdmCallback
+
+            # Tweaked the model from https://github.com/calmisential/TensorFlow2.0_ResNet
+            # MIT License
```

Comment:
Please include their license text at the top of this file below the license of ART.

Reply:
Removing this license as using art architecture for training now.